[otel-agent] address broken log level assignment

truthbk opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?
This PR ensures that:
1) Log levels from the actual environment or datadog.yaml are accounted for.
2) Sets the most verbose level from `DD_LOG_LEVEL` (source via environment or `datadog.yaml`) or the corresponding OTel configuration telemetry logging config defined.

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->
Impossible to set log level with our usual ops, and unexpected behavior when changing the agent logging level to a more verbose mode for troubleshooting.

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->
Modifying `DD_LOG_LEVEL` or `datadog.yaml` `log_level` settings should now be accounted for and applied if more verbose than the OTel telemetry debug verbosity (info by default) - it used to be ignored.

truthbk committed:
[otel-agent] address broken log level assignment

jackgopack4 commented on cmd/otel-agent/config/agent_config.go:
```diff
@@ -75,6 +105,22 @@ func NewConfigComponent(ctx context.Context, ddCfg string, uris []string) (confi
+			logLevel = telemetryLogLevel.String()
+		}
+	}
+	pkgconfig.Set("log_level", logLevel, pkgconfigmodel.SourceLocalConfigProcess)
```
is there ever a situation where logLevelMap wouldn't have a log_level value AND logLevelMap wouldn't also have a valid telemetryLogLevel value? is default loglevel setting handled elsewhere?

jackgopack4 commented:
is this something that requires a release note? I'd be happy to take a stab at writing it if necessary. Other than that I am still curious about the edge case where log level is not set in the Datadog or OTel config, but I'm guessing it gets validated before ever getting to this point? Thanks

songy23 commented on cmd/otel-agent/config/agent_config.go:
```diff
@@ -25,6 +25,36 @@ import (
+
+type level struct {
+	value   logLevel
+	mapping string
```
Is this field ever used (except when this struct is initialized)?

truthbk replied:
It used to, but we can clean it up, not much value in keeping it around.

songy23 commented on cmd/otel-agent/config/agent_config.go:
```diff
@@ -75,6 +105,22 @@ func NewConfigComponent(ctx context.Context, ddCfg string, uris []string) (confi
+	agentLogLevel := pkgconfig.GetString("log_level")
+	telemetryLogLevel := sc.Telemetry.Logs.Level
+	agentLogMapping, ok := logLevelMap[agentLogLevel]
+	if ok {
```
I would rather return an error here since I don't think `ok` should ever be false. If it is false, the mapping is incomplete and we should update.

jackgopack4 replied:
thanks, I think this ties into my comment above and is a better approach.

songy23 commented on cmd/otel-agent/config/agent_config.go:
```diff
@@ -75,6 +105,22 @@ func NewConfigComponent(ctx context.Context, ddCfg string, uris []string) (confi
+		logLevel = agentLogLevel
+	}
+	telemetryLogMapping, ok := logLevelMap[telemetryLogLevel.String()]
+	if ok {
```
Same here

dineshg13 commented on cmd/otel-agent/config/agent_config.go:
```diff
@@ -25,6 +25,36 @@ import (
+
+const (
+	off logLevel = iota
+	critical
+	err
+	warn
+	info
+	debug
+	trace
```
The order seems reverse of general order for levels with ex: critical should have a higher number than info

truthbk replied:
The order is increasing from least verbose to most verbose. Not sure I understand the concern here? This allows us to compare the log values directly.

dineshg13 replied:
See https://github.com/uber-go/zap/blob/f87c7be57ccc7ab7a0aa3fbc12398369fe3a61c3/zapcore/level.go#L37 or https://opentelemetry.io/docs/specs/otel/logs/data-model/#severity-fields
info is generally lower number than critical ?

truthbk committed:
[otel-agent] config: remove mapping + unit test

[otel-agent] config: reset config after each test

[otel-agent] config: run tests in suite

truthbk committed:
[otel-agent] config: feedback - reverse log_level iota

jackgopack4 commented on cmd/otel-agent/config/agent_config.go:
```diff
@@ -25,6 +25,40 @@ import (
+)
+
+// datadog agent log levels: trace, debug, info, warn, error, critical, and off
+// otel log levels: disabled, debug, info, warn, error
```
Will these values/mappings ever be changing over time? If so, will customers need to upgrade their otel agent to get accurate functionality? And if that would be a problem would we have a way to modify these via a config file instead of hard coded

truthbk replied:
The log mapping reflects the agent mapping. It is not expected to change. If new options are added to the mapping they would be very unlikely to result in the removal of any of these from the map (though other new ones could be included). In any case, it would be rare to see these change anyhow.

jackgopack4 commented:
Looks good to me, just was curious about the log level mapping and if we ever expect it to change.

songy23 reviewed (approved):
LGTM

truthbk commented:
/merge
